File: vuln-sandbox/python/cwe-643/xpath-injection.py
# VULNERABLE: InjectXpath — XPath query built with f-string containing user input
# Rule: InjectXpathTemplate | CWE-643 | Severity: HIGH

from lxml import etree
from flask import Flask, request, jsonify

app = Flask(__name__)

# Sample XML document (in production this would be loaded from a file or DB)
XML_DATA = b"""
<users>
  <user id="1"><name>Alice</name><role>admin</role></user>
  <user id="2"><name>Bob</name><role>user</role></user>
  <user id="3"><name>Charlie</name><role>user</role></user>
</users>
"""


@app.route('/api/users/find', methods=['GET'])
def find_user():
    """Find a user by name using XPath."""
    name = request.args.get('name', '')

    tree = etree.fromstring(XML_DATA)

    # <-- VULNERABLE: user input embedded in XPath expression via f-string
    # An attacker can pass: ' or '1'='1 to match all users and bypass access controls
    xpath_query = f"//user[name='{name}']"
    results = tree.xpath(xpath_query)

    users = [{'id': node.get('id'), 'name': node.findtext('name'), 'role': node.findtext('role')}
             for node in results]
    return jsonify({'users': users})


if __name__ == '__main__':
    app.run()

File: vuln-sandbox/node/cwe-643/xpath-injection.js
// VULNERABLE: xpath-injection — XPath query built with string concatenation of user input
// Rule: InjectXpath | CWE-643 | Severity: HIGH

const express = require('express');
const xpath = require('xpath');
const { DOMParser } = require('xmldom');

const app = express();
app.use(express.json());

const xmlDoc = new DOMParser().parseFromString(
  '<users><user><name>admin</name><role>admin</role></user></users>'
);

app.get('/user', (req, res) => {
  const username = req.query.name;

  // VULNERABLE: user input concatenated into XPath expression — allows XPath injection
  const expression = "//user[name='" + username + "']";
  const nodes = xpath.select(expression, xmlDoc);

  res.json({ found: nodes.length > 0 });
});

app.listen(3000);
